### PR TITLE
fix(frontend): guard tokenizer.chat_template for MistralTokenizer

### PR DESCRIPTION
**Pull request by krishung5**
## Overview:

Fix an `AttributeError` that prevents Mistral models from registering on the
vLLM chat-processor path (`--dyn-chat-processor vllm --tokenizer-mode mistral`).

Closes DIS-2626

## Details:

`EngineFactory.chat_engine_factory` in `vllm_processor.py` read
`tokenizer.chat_template` directly. With `--tokenizer-mode mistral`, vLLM returns
a `MistralTokenizer` (`vllm.tokenizers.mistral`) which has **no** `chat_template`
attribute, so the read raised:

```
chat_engine_factory callback failed: AttributeError:
'MistralTokenizer' object has no attribute 'chat_template'
```

The callback failure surfaced as `Error adding model from discovery`, and the
model never registered — every chat request returned HTTP 404.

The fix reads the attribute with `getattr(tokenizer, "chat_template", None)`.
The subsequent assignment still works (the base `TokenizerLike` provides
`__dict__`) and is inert for Mistral, whose rendering is already dispatched by
tokenizer type in `prepost.py` rather than via an HF chat template. The block is
extracted into a small `_ensure_chat_template` helper, and three regression
tests are added exercising the real `MistralTokenizer`.

Verified end-to-end on the vLLM runtime container with
`mistralai/Mistral-7B-Instruct-v0.3` (`--tokenizer-mode mistral`):
- **Before:** frontend `AttributeError`, model does not register, chat → HTTP 404.
- **After:** model registers, `POST /v1/chat/completions` → HTTP 200 with a valid
  completion; zero `chat_template` errors.
- Unit tests: 3 passed.

## Where should the reviewer start?

- `components/src/dynamo/frontend/vllm_processor.py` — the new `_ensure_chat_template`
  helper and its call site in `EngineFactory.chat_engine_factory`.
- `components/src/dynamo/frontend/tests/test_vllm_processor_unit.py` — the three
  added regression tests.

## Related Issues

**🚫 This PR is NOT linked to an issue**:
- [x] Confirmed — no related issue

🤖 Generated with [Claude Code](https://claude.com/claude-code)

---

<a href="https://nvidia.devinenterprise.com/review/ai-dynamo/dynamo/pull/12659" target="_blank">
  <picture>
    <source media="(prefers-color-scheme: dark)" srcset="https://static.devin.ai/assets/gh-open-in-devin-review-dark.svg?v=1">
    Open in Devin Review
  </picture>
</a>

## Summary by CodeRabbit

- **Bug Fixes**
  - Improved compatibility with Mistral tokenizers that do not provide a chat template.
  - Preserved existing Hugging Face tokenizer templates.
  - Ensured explicit chat-template overrides continue to work correctly.
- **Tests**
  - Added regression coverage for tokenizer template handling and safe configuration behavior.

### CODE REVIEW

File: components/src/dynamo/frontend/vllm_processor.py [hunk 1]
```diff
@@ -88,6 +88,20 @@ def _runtime_config_context_length(mdc: ModelDeploymentCard) -> int | None:
...
+) -> None:
+    """Set tokenizer.chat_template so vLLM's renderer handles tool calls.
+
+    Uses getattr, not direct access: MistralTokenizer (--tokenizer-mode mistral)
+    has no chat_template attribute. The assignment still works and is inert there.
+    """
+    if getattr(tokenizer, "chat_template", None) is None:
+        tokenizer.chat_template = resolve_chat_template(local_dir, backend="vllm")
+    if chat_template_flag:
+        tokenizer.chat_template = load_chat_template(chat_template_flag)
```
[the remark this answers is not in the document]
Comment (krishung5) [reply components/src/dynamo/frontend/vllm_processor.py]: The helper now skips the whole block when the tokenizer has no chat_template attribute, so a MistralTokenizer is left entirely untouched rather than getting an inert HF template. This also better matches the original intent — only manage a template for tokenizers that actually have one.

Added test_ensure_chat_template_mistral_ignores_on_disk_template, which writes a chat_template.jinja into the model dir and asserts the Mistral tokenizer is not mutated — covering the non-None on-disk case you flagged (the earlier test only exercised an empty dir).

**Review by rmccorm4 — APPROVED** [review]
